src/metabase/driver.clj

```diff
@@ -940,13 +940,26 @@
   dispatch-on-initialized-driver
   :hierarchy #'hierarchy)
 
-(defmulti truncate!
+(defmulti with-transaction*
+  "Executes the given thunk within a transaction, if the given driver supports transactions."
+  {:added "0.50.0", :arglists '([driver db-id thunk])}
+  dispatch-on-initialized-driver
+  :hierarchy #'hierarchy)
+
+(defmethod with-transaction* ::driver [_ _ thunk]
+  (thunk))
+
+(defmacro with-transaction
+  "Executes the given body within a transaction, if the given driver supports transactions."
+  {:added "0.50.0", :arglists '([driver db-id & body])}
+  [driver db-id & body]
+  `(with-transaction* ~driver ~db-id (fn [] ~@body)))
+
+(defmulti delete!
```

Comment:
Contrary to what an LLM told me, I couldn't get `TRUNCATE` to work reliably within a transaction. Given there were already some semantic mismatches between databases with this command, the camel's back broke and I opted for a more general `DELETE` wrapper.

Supporting the `WHERE` component is left out for now, as I'm not sure how expressive we want it to be, and whether we want to support any other options. Supporting the syntactic flexibility of toucan2 might be nice, but seems non-trivial even if we reuse some of that library's internals.

Reply:
Hmm, even with this change it seems that MariaDB is still hitting [some kind of deadlock.](https://github.com/metabase/metabase/actions/runs/8434588496/job/23105884124?pr=40610)

Reply:
It seems strange to support `delete` without conditions. 

and with our promise to having a 3-deprecation cycle on every method changes I suspect we'll come to a point where we introduce a new `delete-where!` defmulti just to get around that.

How about we take it a step further and implement it now?

Reply:
I believe we could add no-condition delete now and let one of the `& args` be a condition map in 51 without breaking our promises about deprecation.

Reply:
Yeah the idea is that args will leave things open, and we won't need to depreciate. Whether we add a new method or require implementations to handle new parameters, it can still block an upgrade. 

I'm a bit loathe to jump into the details of how we filter, as doing it as generally as toucan does will be complex, and it's unclear how flexibly we'll ever need it to be.

Another option is to keep calling this method truncate. Semantically that's what it's doing, but violates least surprise that it's not how it's implemented.